ops/framework.py

```diff
@@ -146,6 +146,11 @@ def __set_name__(self, emitter_type, event_kind):
     def __get__(self, emitter, emitter_type=None):
         if emitter is None:
             return self
+        # Framework might not be available if accessed as CharmClass.on.event rather than charm_instance.on.event,
+        # but in that case it couldn't be emitted anyway, so there's no point to registering it.
+        framework = getattr(emitter, 'framework', None)
+        if framework is not None:
+            framework.register_type(self.event_type, emitter, self.event_kind)
```

Comment:
So I believe you added this because of your comment:
> Moving the registration of events (the most recent commit) was required to allow dynamic event types to be registered. Previously, when a dynamic event was defined, it relied on the fact that the next time the 

I'm trying to understand what is going on here. The issue is that if we cache EventsBase, then defining a new event type doesn't end up registered?

I definitely like getting rid of the inspect.getmembers() loop in __init__.py since *any* object being initialized then did a search on all attributes, which seems wasteful. (you need to do it 1 time per class, not 1 time per object).

Is there a way to register them when we define the type rather than when we grab one to emit/observe it?

(eg, put it into the EventsBase.__init__ and/or the define_event methods?)

Reply:
Correct.  The calls to `define_event` happen after the first call to `__get__` because it's called as `charm.on.define_event()`, so if the registration happens only on the first `__get__`, then it will not pick up any dynamic events.  It worked previously because the descriptor is then called again on every access, and it works now because the registration is moved to accessing the individual event rather than accessing the collection as a whole.

The reason we can't register them when defined is that the registration requires the actual instance they're being registered against, not just the type they're defined on, so it does in fact need to be 1 time per object.

I think that is due to the registration being used mainly for restoring from snapshots and thus needing to tie to a specific instance, but I haven't looked deep enough into whether there's a way we could register against just the type and still have the instance info we need when snapshotting / restoring.

Reply:
understood